frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -135,26 +135,86 @@ func ListTargets(ctx context.Context, dt []byte) (*targets.List, error) {
 	return l, nil
 }
 
+type namedContextState struct {
```

Comment:
Maybe something like `filteredLocalState` or similar?

I'm not sure about including an `Image` field in here, I get the logic, but it's not really connected with what paths get pulled from the client, it shouldn't affect the contents of the image data. If it gets split out, it feels like a separate refactor IMO, so if there's really a good reason to do this, it makes sense as a separate commit.